### PR TITLE
Add ping-pong (heartbeat) to detect if 1 side of a peer connection closes

### PR DESCRIPTION
**Pull request by dborkan**
Add ping-pong (heartbeat) to detect if 1 side of a peer connection closes.  This fixes issue https://github.com/uProxy/uProxy/issues/151 (so that if a client closes, the server's UI can remove the X is proxying through you text).  This is also necessary to have in place for fixing issue https://github.com/uProxy/uProxy/issues/178 (that will require some new UI)

Tested with manually and with "grunt test"

### CODE REVIEW

File: src/rtc-to-net/rtc-to-net.ts [hunk 1]
```diff
@@ -37,10 +41,13 @@ module RtcToNet {
...
       this.transport = freedom['transport']();
       this.transport.on('onData', this.passPeerDataToNet_);
-      this.transport.on('onClose', this.closeNetClient_);
+      this.transport.on('onClose', this.onCloseHandler_);
       this.transport.setup('RtcToNet-' + peerId, channel.identifier).then(
         // TODO: emit signals when peer-to-peer connections are setup or fail.
-        () => { dbg('RtcToNet transport.setup succeeded'); },
+        () => {
+          dbg('RtcToNet transport.setup succeeded');
+          this.startPingPong_();;
```
Comment (trevj) [inline src/rtc-to-net/rtc-to-net.ts]: typo (two semi-colons)

File: src/socks-to-rtc/socks-to-rtc.ts [hunk 2]
```diff
@@ -217,6 +242,10 @@ module SocksToRTC {
           // Receiving a disconnect on the remote peer should close SOCKS.
           dbg(command.tag + ' <--- received NET-DISCONNECTED');
           this.closeConnectionToPeer(command.tag);
+        } else if (command.type === Channel.COMMANDS.PONG) {
+          // Receiving a disconnect on the remote peer should close SOCKS.
```
Comment (trevj) [inline src/socks-to-rtc/socks-to-rtc.ts]: I think this comment can be removed :-)

File: src/rtc-to-net/rtc-to-net.ts [hunk 3]
```diff
@@ -68,14 +75,16 @@ module RtcToNet {
     /**
      * Close PeerConnection and all TCP sockets.
      */
-    public close = () => {
+    private onCloseHandler_ = () => {
+      dbg('transport closed from peerId ' + this.peerId);
```
Comment (trevj) [inline src/rtc-to-net/rtc-to-net.ts]: Perhaps re-think the wording here? It's not clear to me what "transport closed from x" means!

**Comment by trevj** [comment]
Cool.

Some nits and two broader things to consider:
1. It would be great to have a comment explaining why this is necessary because, as i understand it, WebRTC should have its own ping mechanism (to keep NAT mappings alive, if nothing else). For example, there is a comment explaining the 'start' message which also explains that the 'start' message will eventually go away.
2. If each side will consider the connection a failure unless it receives a ping from the other within some time interval then it seems there's no real need for a ping-pong cycle? If so, the code becomes basically the same on both sides and can probably be pulled out into some common location.

FWIW, I think #1 is more important than #2 :-)

**Comment by dborkan** [comment]
Just added some comments about why we need ping-pong (the short answer is webrtc doesn't notify us if the peer disconnects).

Also the code isn't shared because it is different on the client (socks-to-rtc) and server (rtc-to-net) sides.  For example on the client side we need 2 timers: 1 to send pings every second, another to check for pongs every 10 seconds.  But on the server side, we only need 1 timer to check for pings every 10 seconds, because the pongs are sent back immediately after a ping is received.

**Pull request merged**